### ISSUE TITLE
`read_timed` for Non-Linux (and maybe move to poll())

### ISSUE THREAD

**Issue by domenukk** [AFLplusplus/AFLplusplus#387]
Right now, `read_timed(...)` uses `select` for timeouts.
One important piece is that, for multiple short reads, the timeout is calculated accordingly.
Right now, the code depends on `select` updating the time struct for each call, however, according to @dgmelski in #383 this may not apply to all systems (i.e. non-Linux).

This should probably be addressed.
While we're at it, we might want to consider switching to `poll` as it is, according to docs, faster. I have not done so as `poll` never updates the time struct.

**Comment by vanhauser-thc** [comment]
just read it up. the difference is that linux keeps the time not waited in the time struct, whereas *BSD does not. so for linux it works as expected, for BSD basically an undefined value is returned.
poll needs a little bit more for setup.
I would propose we keep the implementation like it is and just add an #ifdef not linux stuff that ... dunno, real time measurement between the two.
@devnexen ?

**Comment by domenukk** [comment]
Any good ideas, how to get the time measurements quickly?
About `poll`, how do you mean "more for setup"? More code or more time? 
More code might be worth it if it gets faster, actually (and if we need if for BSD anyway).

**Comment by dgmelski** [comment]
`get_cur_time_us` is using `gettimeofday` which uses VDSO to avoid going all the way to the kernel (as I understand it). In contrast, I'd expect the system calls (select, read) and I/O to be much slower. Or perhaps gettimeofday is not sufficiently precise?

**Comment by vanhauser-thc** [comment]
well we need select (or something different, however poll and epoll are not faster, same speed for 1 fd).
because of how select works on linux it does automatically what we want and do not need extra code. for poll and epoll we would need extra code.
therefore IMHO a simply ifdef for not linux that uses get_cur_time before and after the call and returns that keeps linux the (minimal) advantage and work work anywhere else

**Comment by devnexen** [comment]
As a matter of fact indeed poll is not faster albeit it depends on the os but not with just one file descriptor. However I would suggest using this for all oses or still using select but not using the timeout to avoid all the ifdef things magic. What do you think ?

**Comment by domenukk** [comment]
I agree. Not a fan of ifdef if we don't gain anything (with get_cur_time not leading to a syscall) .
If I remember correctly, the Linux docs specifically state poll is faster but I don't know about the internals...

**Comment by vanhauser-thc** [comment]
epoll is the fastest - if you have thousands of file descriptors.
fuzzing is about speed - not only but in this context it is.
its a few instructions but that for every fuzz.
its 2 ifdefs with altogether 3 lines. no thing.

**Comment by dgmelski** [comment]
If you want the total time to stay below timeout_ms, you need to adjust the timeout passed to select() in every iteration of the loop, which I think makes it slightly longer, though still simple enough.

In AFL++'s use of this function, the loop probably never executes more than once.  The bug I reported has likely never occurred, for this reason. You could possibly get by with a less general function (e.g., without the loop), if you wanted to simplify. FWIW.

**Comment by dgmelski** [comment]
Actually, is that read() non-blocking?

**Comment by domenukk** [comment]
Ah... It should be blocking, but at this point should never block, right? Could mean the select time is a bit too long in a short read case, though.

**Comment by dgmelski** [comment]
Won't blocking IO usually block until all of the requested blocks are available? It means the following is possible:
1. The select returns true, because some, but not 'len', bytes are available.
2. The read blocks until 'len' bytes are available, exceeding timeout.
3. The loop never has a second iteration (the loop is pointless).

Mind you, this will never happen, because AFL++ is asking for just 4 byte reads, which will always all be available at the same time. So the function works the way AFL needs it to, but it is trying to do something more general, which it does not do.

It's not unreasonable to have the function assume that "all 'len' bytes become available at once." The easiest way forward might include making that a documented, explicit assumption.

In short, the original code was working, the patched code will work, but they are both confusing.

**Comment by vanhauser-thc** [comment]
@dgmelski is right that in an incomplete read the second read would block.
and the timeout would need to be adjusted every time the select is called.

and I agree that an incomplete read should never happen, even if this would be on tcp/ip in the future - as long as this is just these 4 bytes.

so basically overengineered :)

I simplified the function and documented the assumption that the len bytes must be available to read.

**Comment by vanhauser-thc** [comment]
now looks like this:
```
/* Wrapper for select() and read(), reading len bytes.
  Assumes that all bytes are available on read!
  Returns the time passed to read.
  If the wait times out, returns timeout_ms + 1;
  Returns 0 if an error occurred (fd closed, signal, ...); */
u32 read_timed(s32 fd, void *buf, size_t len, u32 timeout_ms,
               volatile u8 *stop_soon_p) {

  fd_set         readfds;
  FD_ZERO(&readfds);
  FD_SET(fd, &readfds);
  struct timeval timeout;

  timeout.tv_sec = (timeout_ms / 1000);
  timeout.tv_usec = (timeout_ms % 1000) * 1000;
#if !defined(__linux__)
  u64 read_start = get_cur_time_us();
#endif

  /* set exceptfds as well to return when a child exited/closed the pipe. */
  int sret = select(fd + 1, &readfds, NULL, NULL, &timeout);

  if (!sret) {

    return timeout_ms + 1;

  } else if (sret < 0) {

    return 0;

  }

  ssize_t len_read = read(fd, ((u8 *)buf), len);
  if (len_read < len) { return 0; }

#if defined(__linux__)
  u32 exec_ms =
      MIN(timeout_ms,
          ((u64)timeout_ms - (timeout.tv_sec * 1000 + timeout.tv_usec / 1000)));
#else
  u32 exec_ms = get_cur_time_us() - read_start;
#endif

  // ensure to report 1 ms has passed (0 is an error)
  return exec_ms > 0 ? exec_ms : 1;

}

```
IMHO so two little ifdefs are fine, as as cheap get_cur_time_us is, it is still an unnecessary overhead on Linux which is by far the most used platform.
(opposing opinions?)

**Comment by domenukk** [comment]
Let's leave the potential loop here please, and make the read non-blocking instead(?) 
I have plans to use it for other things in the future.
The ifdef is fine, I don't have a strong opinion on it.

**Comment by domenukk** [comment]
How about this?
https://github.com/AFLplusplus/AFLplusplus/commit/38e5c32a55086d36c8b9ee38e4b20c15517fc4b2
Not that I opted for the `ioctl` to set nonblocking mode, as it is only one syscall and appears to be widely supported.
It had no measurable performance impact on my machine, however feel free to benchmark and potentially add a `read_timed_4` method for the 4 byte special case, based on the function by @vanhauser-thc

**Comment by domenukk** [comment]
After discussing with @vanhauser-thc again, we don't need a general `read_timed` function as we always read 4 bytes, so I removed it again. New solution here: 2450819
